ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,382 @@
+(* Copyright (C) Cloud Software Group, Inc.
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU Lesser General Public License as published
+   by the Free Software Foundation; version 2.1 only. with the special
+   exception on linking described in file LICENSE.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU Lesser General Public License for more details.
+ *)
+
+let ( // ) = Filename.concat
+
+let runtime_data = "/var/lib" // "xapi-guard"
+
+let ( let* ) = Lwt.bind
+
+let ( let@ ) f x = f x
+
+let with_lock = Lwt_mutex.with_lock
+
+type t = Uuidm.t * Mtime.t * Types.Tpm.key
+
+let cache_of service = runtime_data // Types.Service.to_string service
+
+let files_in dir ~otherwise =
+  Lwt.catch
+    (fun () ->
+      let* listing = Lwt_unix.files_of_directory dir |> Lwt_stream.to_list in
+      List.filter_map
+        (function "." | ".." -> None | name -> Some (dir // name))
+        listing
+      |> Lwt.return
+    )
+    otherwise
+
+let unlink_safe file =
+  let __FUN = __FUNCTION__ in
+  Lwt.catch
+    (fun () -> Lwt_unix.unlink file)
+    (function
+      | Unix.(Unix_error (ENOENT, _, _)) ->
+          Lwt.pause ()
+      | e ->
+          Logs_lwt.info (fun m ->
+              m "%s: error %s when deleting %s, ignoring" __FUN
+                (Printexc.to_string e) file
+          )
+      )
+
+(** - Direct: request doesn't pass through the cache
+    - Engaged: both side coordinate through the queue, writer ends the mode
+      when the queue has been filled.
+    - Disengaged: writer ignores the queue, reader empties it and the cache;
+      then it changes the mode to engaged.
+*)
+type state = Direct | Engaged | Disengaged
+
+type channel = {
+    queue: t Lwt_bounded_stream.t
+  ; push: t option -> unit option
+  ; lock: Lwt_mutex.t (* lock for the states *)
+  ; mutable state: state
+}
+
+(*
+   Notes:
+   - uses Mtime.t to force usage of monotonic time
+   - This means that between runs (and reboots) cached stated is lost if not
+     persisted first.
+   IDEA: carryover: read contents of cache and "convert it" to the current run
+
+   TODO:
+     - Add startup step to convert existing content to new time
+     - Exponential backoff on xapi push error
+     - Limit error logging on xapi push error: once per downtime is enough
+ *)
+
+module Writer : sig
+  val with_cache :
+       direct:(t -> string -> (unit, exn) Lwt_result.t)
+    -> Types.Service.t
+    -> channel
+    -> ((t -> string -> unit Lwt.t) -> 'a Lwt.t)
+    -> 'a Lwt.t
+  (** [with_cache ~direct typ context] creates a cache for content of type [typ]. The
```

Comment:
`with_cache ~direct typ chan` will return a function that can write to the buffer when given a write function `(t -> string -> unit Lwt.t) -> 'a Lwt.t`? Looks like you don't have `channel` in your argument?